File: day16/python/day16.py
from enum import Enum
import os
import sys

class Direction(Enum):
    LEFT = 0,
    RIGHT = 1,
    UP = 2,
    DOWN = 3

def get_reflect_direction(mirror, incoming_direction):
    if(mirror == '\\'):
        if incoming_direction == Direction.LEFT:
            return Direction.UP
        if incoming_direction == Direction.RIGHT:
            return Direction.DOWN
        if incoming_direction == Direction.UP:
            return Direction.LEFT
        if incoming_direction == Direction.DOWN:
            return Direction.RIGHT
    elif( mirror == '/'):
        if incoming_direction == Direction.LEFT:
            return Direction.DOWN
        if incoming_direction == Direction.RIGHT:
            return Direction.UP
        if incoming_direction == Direction.UP:
            return Direction.RIGHT
        if incoming_direction == Direction.DOWN:
            return Direction.LEFT

def get_new_posiiton(row, col, direction):
    if(direction == Direction.LEFT):
        row = row
        col = col - 1
    elif(direction == Direction.RIGHT):
        row = row
        col = col + 1
    elif(direction == Direction.UP):
        row = row - 1
        col = col
    elif(direction == Direction.DOWN):
        row = row + 1
        col = col

    return (row,col)

def solve_rec(current_row, current_col, beam_direction, lines, seen_beam_direction_map):
    rows = len(lines)
    cols = len(lines[0])

    # input args are current position and beam pointing direction.
    # need to work out the next coordinate and what direction beam will face when there

    if(beam_direction in seen_beam_direction_map[current_row][current_col]):
        # a beam on a square will ALWAYS follow the same path it did the first time if travelling same direction
        # so we can bail immediately if we have landed on this square, from this direction before
        return

    # store the direction we came in to this square
    seen_beam_direction_map[current_row][current_col].append(beam_direction)

    # must be a list as sometimes we split the beam and there are two to process
    beams_to_process = []

    if(lines[current_row][current_col] == '.'):
        updated_row, updated_col = get_new_posiiton(current_row, current_col, beam_direction)
        beams_to_process.append((updated_row,updated_col,beam_direction))
        # check if we hit a wall
    elif(lines[current_row][current_col] in "\\/"):
        new_beam_direction = get_reflect_direction(lines[current_row][current_col], beam_direction)
        updated_row, updated_col = get_new_posiiton(current_row, current_col, new_beam_direction)
        beams_to_process.append((updated_row,updated_col,new_beam_direction))
    elif lines[current_row][current_col] in "-|":
        if(lines[current_row][current_col] == '-' and beam_direction in [Direction.LEFT, Direction.RIGHT]):
            # straight through '-'
            updated_row, updated_col = get_new_posiiton(current_row, current_col, beam_direction)
            beams_to_process.append((updated_row,updated_col,beam_direction))
        elif(lines[current_row][current_col] == '|' and beam_direction in [Direction.UP, Direction.DOWN]):
            # straight through '|'
            updated_row, updated_col = get_new_posiiton(current_row, current_col, beam_direction)
            beams_to_process.append((updated_row,updated_col,beam_direction))
        elif(lines[current_row][current_col] == '-' and beam_direction in [Direction.UP, Direction.DOWN]):
            # hitting '-' side-on
            # doesnt matter what incoming direction was because outgoing is alway left AND right
            updated_left_row, updated_left_col = get_new_posiiton(current_row, current_col, Direction.LEFT)
            updated_right_row, updated_right_col = get_new_posiiton(current_row, current_col, Direction.RIGHT)

            beams_to_process.append((updated_left_row,updated_left_col,Direction.LEFT))
            beams_to_process.append((updated_right_row,updated_right_col,Direction.RIGHT))
        elif(lines[current_row][current_col] == '|' and beam_direction in [Direction.LEFT, Direction.RIGHT]):
            # hitting '|' side-on
            # doesnt matter what incoming direction was because outgoing is always up AND DOWN
            updated_up_row, updated_up_col = get_new_posiiton(current_row, current_col, Direction.UP)
            updated_down_row, updated_down_col = get_new_posiiton(current_row, current_col, Direction.DOWN)

            beams_to_process.append((updated_up_row,updated_up_col,Direction.UP))
            beams_to_process.append((updated_down_row,updated_down_col,Direction.DOWN))

    for beam_def in beams_to_process:
        # check for hitting a wall
        if (beam_def[1] < 0 or beam_def[1] >= cols or beam_def[0] < 0 or beam_def[0] >= rows):
            pass # new location would be outside the bounds. end current beam here
        else:
            # didnt hit anything, can continue
            solve_rec(beam_def[0], beam_def[1], beam_def[2], lines, seen_beam_direction_map)

def build_blank_seen_transitions_map(row_count, col_count):
    seen_transitions_map = []
    for i in range(0,row_count):
        seen_transitions_map.append([])
        for j in range(0,col_count):
            seen_transitions_map[i].append([])
    return seen_transitions_map

def count_energized_in_seen_transitions_map(seen_transitions_map):
    count = 0
    for seen_map_row in seen_transitions_map:
        for cell in seen_map_row:
            if cell != []:
                count+=1
    return count

def main():
    part_two = True
    # need slightly more recursion limit than standard
    sys.setrecursionlimit(10000)

    cwd = os.path.dirname(__file__)
    with(open(f"{cwd}/../input.txt", 'r') as fp):
        all_lines = fp.readlines()
    all_lines_trimmed = [x.strip() for x in all_lines]

    starting_positions = []
    row_count = len(all_lines_trimmed)
    col_count = len(all_lines_trimmed[0])

    if(part_two == False):
        starting_positions.append((0,0,Direction.RIGHT))
    else:
        for i in range(0, row_count):
            #add all row-aligned items
            starting_positions.extend([(i, 0, Direction.RIGHT), (i, col_count - 1, Direction.LEFT)])
            
        for i in range(0, col_count):
            #add all col-aligned items
            starting_positions.extend([(0, i, Direction.DOWN), (row_count - 1, i, Direction.UP)])

    best_count = 0
    for starting_position in starting_positions:
        # new blank transition map each time
        seen_transitions_map = build_blank_seen_transitions_map(row_count, col_count)

        # complete the transition map
        solve_rec(starting_position[0], starting_position[1], starting_position[2], all_lines_trimmed, seen_transitions_map)

        # count the non-empty cells in the transition map
        count = count_energized_in_seen_transitions_map(seen_transitions_map)
        
        if(count > best_count):
            best_count = count

    print(f"best count: {best_count}")

if __name__ == "__main__":
    main()

File: day16/c/day16.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include "util.h"

enum DIRECTION {
    LEFT = 0x01,
    RIGHT = 0x02,
    UP = 0x04,
    DOWN = 0x08,
    UNKNOWN = 0x80
};

typedef struct beam_def {
    int16_t row;
    int16_t col;
    enum DIRECTION beam_direction;
    int16_t is_set;
} beam_def;

void set_beam_def(beam_def* beam_def, int16_t row, int16_t col, enum DIRECTION beam_direction){
    beam_def->row = row;
    beam_def->col = col;
    beam_def->beam_direction = beam_direction;
    beam_def->is_set = 1;
}

int16_t get_new_row(int16_t row, enum DIRECTION direction){
    int16_t new_row;
    if(direction == LEFT){
        new_row = row;
    }
    else if(direction == RIGHT){
        new_row = row;
    }
    else if(direction == UP){
        new_row = row - 1;
    }
    else if(direction == DOWN){
        new_row = row + 1;
    }

    return new_row;
}

int16_t get_new_col(int16_t col, enum DIRECTION direction){
    int16_t new_col;
    if(direction == LEFT){
        new_col = col - 1;
    }
    else if(direction == RIGHT){
        new_col = col + 1;
    }
    else if(direction == UP){
        new_col = col;
    }
    else if(direction == DOWN){
        new_col = col;
    }

    return new_col;
}

enum DIRECTION get_reflected_direction(char mirror, enum DIRECTION incoming_direction){
    if( mirror == '\\'){
        switch (incoming_direction) {
            case LEFT:
                return UP;
            case RIGHT:
                return DOWN;
            case UP:
                return LEFT;
            case DOWN:
                return RIGHT;
            default:
                return UNKNOWN;
        }
    }
    else if( mirror == '/'){
        switch (incoming_direction) {
            case LEFT:
                return DOWN;
            case RIGHT:
                return UP;
            case UP:
                return RIGHT;
            case DOWN:
                return LEFT;
            default:
                return UNKNOWN;
        }
    }
}

void pretty_print_transition_map(grid* transition_map){
    for(int i = 0; i< transition_map->row_count; i++){
        for(int j = 0; j < transition_map->col_count; j++){
            if(get_grid_value(transition_map, i, j) == 0){
                printf(".");
            }
            else{
                printf("#");
            }
        }
        printf("\n");
    }
}

void solve_rec(
    grid* map_grid, 
    grid* seen_transition_map, 
    int16_t current_row, 
    int16_t current_col, 
    enum DIRECTION beam_direction){
    char transition_values_for_cell = get_grid_value(seen_transition_map, current_row, current_col);
    if ( (transition_values_for_cell & beam_direction) != 0){
        // logical and of these values will be > 0 if that bit is set
        // immediately return
        return;
    }
    // set the correct bit in that cell to indicate we have been through here in this direction
    
    set_grid_value(seen_transition_map, current_row, current_col, (transition_values_for_cell | beam_direction));

    // number of next beams to recursively process can only be 1 or 2.
    // initialize both to "NULL"
    beam_def next_one = { .is_set = 0 };
    beam_def next_two = { .is_set = 0 };

    char current_position_symbol = get_grid_value(map_grid, current_row, current_col);

    if(current_position_symbol == '.'){
        // pass straight through '.' without changing direction
        int16_t new_row = get_new_row(current_row, beam_direction);
        int16_t new_col = get_new_col(current_col, beam_direction);
        set_beam_def(&next_one, new_row, new_col, beam_direction);
    }
    else if(current_position_symbol == '/' || 
            current_position_symbol == '\\'){
        // for mirrors, calculate the new direction, then move 1 in that direction
        enum DIRECTION new_beam_direction = get_reflected_direction(current_position_symbol, beam_direction);
        int16_t new_row = get_new_row(current_row, new_beam_direction);
        int16_t new_col = get_new_col(current_col, new_beam_direction);
        set_beam_def(&next_one, new_row, new_col, new_beam_direction);
    }
    else if(current_position_symbol == '-'){
        if(beam_direction == LEFT || beam_direction == RIGHT){
            // straight through
            int16_t new_row = get_new_row(current_row, beam_direction);
            int16_t new_col = get_new_col(current_col, beam_direction);
            set_beam_def(&next_one, new_row, new_col, beam_direction);
        }
        else{
            //split. send one left and send one right
            int16_t new_row_l = get_new_row(current_row, LEFT);
            int16_t new_col_l = get_new_col(current_col, LEFT);
            set_beam_def(&next_one, new_row_l, new_col_l, LEFT);

            int16_t new_row_r = get_new_row(current_row, RIGHT);
            int16_t new_col_r = get_new_col(current_col, RIGHT);
            set_beam_def(&next_two, new_row_r, new_col_r, RIGHT);
        }
    }
    else if(current_position_symbol == '|'){
        if(beam_direction == UP || beam_direction == DOWN){
            // straight through
            int16_t new_row = get_new_row(current_row, beam_direction);
            int16_t new_col = get_new_col(current_col, beam_direction);
            set_beam_def(&next_one, new_row, new_col, beam_direction);
        }
        else{
            //split. send one up and send one down
            int16_t new_row_u = get_new_row(current_row, UP);
            int16_t new_col_u = get_new_col(current_col, UP);
            set_beam_def(&next_one, new_row_u, new_col_u, UP);

            int16_t new_row_d = get_new_row(current_row, DOWN);
            int16_t new_col_d = get_new_col(current_col, DOWN);
            set_beam_def(&next_two, new_row_d, new_col_d, DOWN);
        }
    }

    // for each beam that is set
    // check it is in bounds
    // if it is, call this func again with the new values
    if(next_one.is_set == 1){
        if(
            next_one.row >= 0 && 
            next_one.row < map_grid->row_count && 
            next_one.col >= 0 && 
            next_one.col < map_grid->col_count){
            // is in-bounds.
            solve_rec(map_grid, seen_transition_map, next_one.row, next_one.col, next_one.beam_direction);
        }
    }

    if(next_two.is_set == 1){
        if(
            next_two.row >= 0 && 
            next_two.row < map_grid->row_count && 
            next_two.col >= 0 && 
            next_two.col < map_grid->col_count){
            // is in-bounds
            solve_rec(map_grid, seen_transition_map, next_two.row, next_two.col, next_two.beam_direction);
        }
    }
}

int main(int argc, char** argv){
    char* input_file;
    if (argc == 2) {
        input_file = argv[1];
    } else {
        input_file = "../input_short.txt";
    }

    FILE* fp = fopen(input_file, "r");
    int16_t row_count = get_file_row_count(fp);
    int16_t col_count = get_file_col_count(fp);
    grid* map_grid = read_input_grid(fp, row_count, col_count);
    fclose(fp);

    printf("grid rows -> %d\n", map_grid->row_count);
    printf("grid cols -> %d\n", map_grid->col_count);
    pretty_print_grid(map_grid);
    printf("=============== START =================\n");

    grid* seen_transition_map = create_grid(row_count, col_count);

    solve_rec(map_grid, seen_transition_map, 0, 0, RIGHT);

    int count = 0;
    for(int i = 0; i < (row_count * col_count); i++){
        if(seen_transition_map->grid_data[i] != 0){
            count++;
        }
    }

    printf("Count: %d\n", count);
    return 0;
}